Packs/DuoAdminApi/Integrations/DuoEventCollector/DuoEventCollector.py

```diff
@@ -163,14 +230,81 @@ def rotate_request_order(self) -> None:
+            d. all events are in the fetch window (if 1)
+
+            ifs:
+            1.  There isn't a fetch delay or All the events are in the fetch window.
+            2.  The events in the response aren't in the fetch_window
+            3.  The mid event is in the window and the next isn't.
+            4.  The mid-1 event is in the window and the next isn't.
+            5.  The mid event is lower than the end of the window.
+            6.  The mid event is higher than the end of the window.
+
+            Args:
+                events (list[dict]): List of events from the current fetch response.
+
+            Returns:
+                tuple[list[dict], bool]: The list of events, bool represents whether we reached the end of the fetch window.
+            """
+        # if 1
+        if (self.client.params.fetch_delay == '0'
+                or datetime.fromtimestamp(events[-1]['timestamp']) < self.client.params.end_window):
+            demisto.debug(f"events_in_window, all events in the fetch window {events[-1]['timestamp']=} < "
+                          f"{self.client.params.end_window.timestamp()=}")
+            return events, False
+        # if 2
+        if datetime.fromtimestamp(events[0]['timestamp']) >= self.client.params.end_window:
+            demisto.debug(f"events_in_window, no events are in the fetch window {events[0]['timestamp']=} >= "
+                          f"{self.client.params.end_window.timestamp()=}")
+            return [], True
+        low = 0
+        high = len(events) - 1
+        mid = 0
+
+        while low <= high:
+
+            mid = (high + low) // 2
+            # if 3
+            if events[mid]['timestamp'] < self.client.params.end_window.timestamp() <= events[mid + 1]['timestamp']:
+                # this is the last event in the window, return the events up to this event including the event.
+                demisto.debug(f"events_in_window, {events[mid]['timestamp']=} < {self.client.params.end_window.timestamp()=} "
+                              f"<= {events[mid + 1]['timestamp']=}, {mid=}")
+                return events[:mid + 1], True
+            # if 4
+            elif events[mid - 1]['timestamp'] < self.client.params.end_window.timestamp() <= events[mid]['timestamp']:
+                demisto.debug(f"events_in_window, {events[mid - 1]['timestamp']=} < "
+                              f"{self.client.params.end_window.timestamp()=} <= {events[mid]['timestamp']=}, {mid=}")
+                return events[:mid], True
+            # if 5
+            elif events[mid]['timestamp'] < self.client.params.end_window.timestamp():
+                demisto.debug(f"events_in_window, {events[mid]['timestamp']} < {self.client.params.end_window.timestamp()}, "
+                              f"{mid=}")
+                low = mid + 1
+            # if 6
+            else:  # events[mid]['timestamp'] > self.client.params.end_window.timestamp():
```

Comment:
This is not the only case where the other statement will occur. It will happen also when
events[mid]['timestamp'] == self.client.params.end_window.timestamp()

And we might need to think of that case. In that case, we will set the high to be mid-1, and then the new mid is supposed to be lower than the end window - this should be ok. just validate edge cases like this.

Reply:
The case of `events[mid]['timestamp'] == self.client.params.end_window.timestamp()` is covered by if no.4 (`elif events[mid - 1]['timestamp'] < self.client.params.end_window.timestamp() <= events[mid]['timestamp']:`.